### PR TITLE
KIP 951 - Phase 1

### PR DESCRIPTION
**Pull request by anchitj**
Taken common structs from https://github.com/confluentinc/librdkafka/pull/4609/files.

### CODE REVIEW

File: src/rdkafka_request.c [hunk 1]
```diff
@@ -460,6 +460,59 @@ int rd_kafka_buf_write_topic_partitions(
 }
 
 
+/**
+ * @brief Read current leader from \p rkbuf.
+ *
+ * @param rkbuf buffer to read from
+ * @param CurrentLeader is the CurrentLeader to populate.
+ * @returns 1 on success, else -1 on parse error.
```
Comment (emasab) [inline src/rdkafka_request.c:468]: better standardizing on return as returns is an alias for that
```suggestion
 * @return 1 on success, else -1 on parse error.
```

**Review by emasab — CHANGES_REQUESTED** [review]
LGTM, just so minor changes

File: src/rdkafka_buf.h [hunk 2]
```diff
@@ -820,6 +820,29 @@ struct rd_kafka_buf_s { /* rd_kafka_buf_t */
                 }                                                              \
         } while (0)
 
+/**
+ * @brief Read KIP-482 Tags at the current position in the buffer using
```
[the remark this answers is not in the document]
Comment (anchitj) [reply src/rdkafka_buf.h]: Updated

File: src/rdkafka_request.c (hunk 1 above)
Comment (anchitj) [reply src/rdkafka_request.c]: Updated.

**Comment by anchitj** [comment]
Rebased ontop of https://github.com/confluentinc/librdkafka/pull/4584

File: src/rdkafka_request.c [hunk 3]
```diff
@@ -497,6 +497,59 @@ int rd_kafka_buf_write_topic_partitions(
...
+ * @param NodeEndpoints is the NodeEndpoints to populate.
+ * @returns 1 on success, else -1 on parse error.
+ */
+int rd_kafka_buf_read_NodeEndpoints(rd_kafka_buf_t *rkbuf,
+                                    rd_kafkap_NodeEndpoints_t *NodeEndpoints) {
+        const int log_decode_errors = LOG_ERR;
+        int32_t i;
+        rd_kafka_buf_read_arraycnt(rkbuf, &NodeEndpoints->NodeEndpointCnt,
+                                   RD_KAFKAP_BROKERS_MAX);
+        RD_IF_FREE(NodeEndpoints->NodeEndpoints, rd_free);
```
Comment (emasab) [inline src/rdkafka_request.c:531]: This should be called only once, so you can replace it with an assert
```suggestion
        rd_dassert(!NodeEndpoints->NodeEndpoints);
```

File: src/rdkafka_proto.h [hunk 4]
```diff
@@ -726,5 +726,31 @@ rd_kafka_pid_bump(const rd_kafka_pid_t old) {
 
 /**@}*/
 
+/**
```
Comment (emasab) [inline src/rdkafka_proto.h:729]: Please move these structs to `rdkafka_request.h` after `rd_kafka_topic_partition_field_t` where we'll keep the Produce and Fetch tags structures too

File: src/rdkafka_request.c [hunk 5]
```diff
@@ -497,6 +497,59 @@ int rd_kafka_buf_write_topic_partitions(
 }
 
 
+/**
+ * @brief Read current leader from \p rkbuf.
+ *
+ * @param rkbuf buffer to read from
+ * @param CurrentLeader is the CurrentLeader to populate.
+ * @return 1 on success, else -1 on parse error.
```
Comment (emasab) [inline src/rdkafka_request.c:505]: add a newline to separate it from the parameters

File: src/rdkafka_request.c [hunk 6]
```diff
@@ -497,6 +497,59 @@ int rd_kafka_buf_write_topic_partitions(
...
+err_parse:
+        return -1;
+}
+
+/**
+ * @brief Read NodeEndpoints from \p rkbuf.
+ *
+ * @param rkbuf buffer to read from
+ * @param NodeEndpoints is the NodeEndpoints to populate.
+ * @returns 1 on success, else -1 on parse error.
```
Comment (emasab) [inline src/rdkafka_request.c:523]: - add a newline to separate it from the parameters
- use return as returns is the alias

File: src/rdkafka_buf.h [hunk 7]
```diff
@@ -836,6 +836,29 @@ struct rd_kafka_buf_s { /* rd_kafka_buf_t */
                 }                                                              \
         } while (0)
 
+/**
+ * @brief Read KIP-482 Tags at current position in the buffer using
+ *        the `read_tag` function receiving the `opaque' pointer.
+ */
+#define rd_kafka_buf_read_tags(rkbuf, read_tag, opaque)                        \
```
Comment (emasab) [inline src/rdkafka_buf.h:843]: Change opaque to `...` for flexibility
```suggestion
#define rd_kafka_buf_read_tags(rkbuf, read_tag, ...)                        \
```

File: src/rdkafka_buf.h [hunk 8]
```diff
@@ -836,6 +836,29 @@ struct rd_kafka_buf_s { /* rd_kafka_buf_t */
...
+                uint64_t _tagcnt;                                              \
+                if (!((rkbuf)->rkbuf_flags & RD_KAFKA_OP_F_FLEXVER))           \
+                        break;                                                 \
+                rd_kafka_buf_read_uvarint(rkbuf, &_tagcnt);                    \
+                while (_tagcnt-- > 0) {                                        \
+                        uint64_t _tagtype, _taglen;                            \
+                        rd_kafka_buf_read_uvarint(rkbuf, &_tagtype);           \
+                        rd_kafka_buf_read_uvarint(rkbuf, &_taglen);            \
+                        int _read_tag_resp =                                   \
+                            read_tag(rkbuf, _tagtype, _taglen, opaque);        \
```
Comment (emasab) [inline src/rdkafka_buf.h:854]:
```suggestion
                            read_tag(rkbuf, _tagtype, _taglen, opaque);        \
```

File: src/rdkafka_buf.h [hunk 9]
```diff
@@ -836,6 +836,29 @@ struct rd_kafka_buf_s { /* rd_kafka_buf_t */
...
+                        if (_read_tag_resp == -1)                              \
+                                goto err_parse;                                \
+                        if (!_read_tag_resp && _taglen > 0)                    \
+                                rd_kafka_buf_skip(rkbuf, (size_t)(_taglen));   \
+                }                                                              \
+        } while (0)
+
```
Comment (emasab) [inline src/rdkafka_buf.h:861]: Add the generic define for writing tags too here.

```suggestion

/**
 * @brief Write \p tagcnt tags at the current position in the buffer.
 * Calling \p write_tag to write each one with \p rkbuf , tagtype
 * argument and the remaining arguments.
 */
#define rd_kafka_buf_write_tags(rkbuf, write_tag, tags, tagcnt, ...)           \
        do {                                                                   \
                uint64_t i;                                                    \
                if (!((rkbuf)->rkbuf_flags & RD_KAFKA_OP_F_FLEXVER))           \
                        break;                                                 \
                rd_kafka_buf_write_uvarint(rkbuf, tagcnt);                     \
                for (i = 0; i < tagcnt; i++) {                                 \
                        size_t of_taglen, prev_buf_len;                        \
                        rd_kafka_buf_write_uvarint(rkbuf, tags[i]);            \
                        of_taglen    = rd_kafka_buf_write_arraycnt_pos(rkbuf); \
                        prev_buf_len = (rkbuf)->rkbuf_buf.rbuf_len;            \
                        write_tag(rkbuf, tags[i], __VA_ARGS__);                \
                        rd_kafka_buf_finalize_arraycnt(                        \
                            rkbuf, of_taglen,                                  \
                            (rkbuf)->rkbuf_buf.rbuf_len - prev_buf_len - 1);   \
                }                                                              \
        } while (0)
```

**Review by emasab — CHANGES_REQUESTED** [review]
Some changes that will be needed for next commits

File: src/rdkafka_buf.h (hunk 8 above)
Comment (anchitj) [reply src/rdkafka_buf.h:854]: It's the same

File: src/rdkafka_request.c (hunk 3 above)
Comment (anchitj) [reply src/rdkafka_request.c]: Updated

File: src/rdkafka_request.c (hunk 5 above)
Comment (anchitj) [reply src/rdkafka_request.c:506]: Updated

File: src/rdkafka_request.c (hunk 6 above)
Comment (anchitj) [reply src/rdkafka_request.c]: Updated

File: src/rdkafka_buf.h (hunk 7 above)
Comment (anchitj) [reply src/rdkafka_buf.h]: Updated

File: src/rdkafka_buf.h (hunk 9 above)
Comment (anchitj) [reply src/rdkafka_buf.h:861]: Updated

File: src/rdkafka_buf.h (hunk 8 above)
Comment (emasab) [reply src/rdkafka_buf.h]: Yeah, meant `__VA_ARGS__`

**Review by emasab — CHANGES_REQUESTED** [review]
Last few changes

File: src/rdkafka_proto.h [hunk 10]
```diff
@@ -726,5 +726,4 @@ rd_kafka_pid_bump(const rd_kafka_pid_t old) {
 
 /**@}*/
 
-
```
[the remark this answers is not in the document]
Comment (anchitj) [reply src/rdkafka_proto.h]: Updated

File: src/rdkafka_buf.h [hunk 11]
```diff
@@ -836,6 +836,53 @@ struct rd_kafka_buf_s { /* rd_kafka_buf_t */
                 }                                                              \
         } while (0)
 
+/**
+ * @brief Read KIP-482 Tags at current position in the buffer using
+ *        the `read_tag` function receiving the `opaque' pointer.
+ */
+#define rd_kafka_buf_read_tags(rkbuf, read_tag, ...)                        \
```
[the remark this answers is not in the document]
Comment (anchitj) [reply src/rdkafka_buf.h]: Updated

File: src/rdkafka_request.h [hunk 12]
```diff
@@ -83,6 +83,28 @@ typedef enum {
         RD_KAFKA_TOPIC_PARTITION_FIELD_NOOP,
 } rd_kafka_topic_partition_field_t;
 
+/**
+ * @name Current Leader and NodeEndpoints for KIP-951
```
[the remark this answers is not in the document]
Comment (anchitj) [reply src/rdkafka_request.h:87]: Updated

**Review by emasab — APPROVED** [review]
LGTM

**Pull request merged**